prog/vm/github_runner.rb

```diff
@@ -0,0 +1,162 @@
+    hop_wait_bootstrap_rhizome
+  end
+
+  label def wait_bootstrap_rhizome
+    reap
+    hop_install_actions_runner if leaf?
+    donate
+  end
+
+  # TODO: Move this to golden image too
+  label def install_actions_runner
+    vm.sshable.cmd("curl -o actions-runner-linux-x64-2.308.0.tar.gz -L https://github.com/actions/runner/releases/download/v2.308.0/actions-runner-linux-x64-2.308.0.tar.gz")
+    vm.sshable.cmd("echo '9f994158d49c5af39f57a65bf1438cbae4968aec1e4fec132dd7992ad57c74fa  actions-runner-linux-x64-2.308.0.tar.gz' | shasum -a 256 -c")
+    vm.sshable.cmd("tar xzf ./actions-runner-linux-x64-2.308.0.tar.gz")
+
+    hop_register_runner
+  end
+
+  label def register_runner
+    unless github_runner.runner_id
+      # We use generate-jitconfig instead of registration-token because it's
+      # recommended by GitHub for security reasons.
+      # https://docs.github.com/en/actions/security-guides/security-hardening-for-github-actions#using-just-in-time-runners
+      data = {name: github_runner.ubid.to_s, labels: ["ubicloud"], runner_group_id: 1}
+      response = github_client.post("/repos/#{github_runner.repository_name}/actions/runners/generate-jitconfig", data)
+      github_runner.update(runner_id: response[:runner][:id], ready_at: Time.now)
+      vm.sshable.cmd("common/bin/daemonizer 'sudo -u runner /home/runner/run.sh --jitconfig #{response[:encoded_jit_config].shellescape}' runner-script")
+    end
+
+    case vm.sshable.cmd("common/bin/daemonizer --check runner-script")
+    when "Succeeded", "InProgress"
+      hop_wait
+    when "NotStarted", "Failed"
+      github_client.delete("/repos/#{github_runner.repository_name}/actions/runners/#{github_runner.runner_id}")
+      github_runner.update(runner_id: nil, ready_at: nil)
+    end
+    nap 10
+  end
+
+  label def wait
+    # If the runner doesn't pick a job in two minutes, destroy it
+    if github_runner.job_id.nil? && Time.now > github_runner.ready_at + 60 * 2
+      response = github_client.get("/repos/#{github_runner.repository_name}/actions/runners/#{github_runner.runner_id}")
+      unless response[:busy]
+        github_runner.incr_destroy
+        nap 0
+      end
+    end
+
+    if vm.sshable.cmd("common/bin/daemonizer --check runner-script") == "Succeeded"
+      github_runner.incr_destroy
+      nap 0
+    end
+
+    nap 15
+  end
+
+  label def destroy
+    register_deadline(nil, 10 * 60)
+
+    decr_destroy
+
+    hop_check_runner_deregistered
+  end
+
+  label def check_runner_deregistered
+    github_client.get("/repos/#{github_runner.repository_name}/actions/runners/#{github_runner.runner_id}")
+    github_client.delete("/repos/#{github_runner.repository_name}/actions/runners/#{github_runner.runner_id}")
+    nap 5
+  rescue Octokit::NotFound
+    hop_destroy_models
+  end
+
+  label def destroy_models
+    vm.private_subnets.each { _1.incr_destroy }
+    vm.sshable.destroy
+    vm.incr_destroy
+    github_runner.destroy
+
+    pop "github runner deleted"
```

Comment:
I added `wait_vm_destroy` label. I added check for vm, not for private subnet.

We access private subnet via vm. But after vm is destroyed, we don't have reference to private subnet. In addition, private subnet can't destroyed before vm. So it will be destroyed after vm.

I don't want to make it more complex. I think waiting for vm destroy is enough for now.